# Problem Set 9
## Juan Roa A15541981

### 1st block (imports)

In [1]:
import seaborn as sns
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression

data = sns.load_dataset('geyser')

### 2nd block (k-fold)

In [57]:
# creates sections for testing data
start = list(np.arange(0,272, 27))
stop = list(np.arange(27,272,27))

# defines folds for kfold
folds = 10

# empty list for R2 values 
R2_scores = []

# kfold process
for i in range(folds):
  # creates testing and training data 
  testing = data[start[i]:stop[i]]
  training = data[~data.isin(testing)].dropna()

  # creates testing x and y values 
  X_test = testing.iloc[:,0]
  y_test = testing.iloc[:,1]

  # reshapes X_test into a 2d array
  X_test = np.array(X_test)
  X_test = X_test.reshape(len(X_test), 1)

  # creates training x and y values
  X_train = training.iloc[:,0]
  y_train = training.iloc[:,1]

  # reshapes X_train into a 2d array
  X_train = np.array(X_train)
  X_train = X_train.reshape(len(X_train), 1)


  # creates linear regression model
  model = LinearRegression()
  model.fit(X_train, y_train)

  # appends R2 scores to list
  R2_scores.append(model.score(X_test, y_test))

print(R2_scores)

[0.8329032381487851, 0.7710004264612066, 0.6209837424442065, 0.8499486244295622, 0.8418983776052437, 0.803328491445922, 0.8393558600070665, 0.7655063509953643, 0.8600645223743294, 0.7787918544568682]


### 3rd block (average and std for each fold)

In [59]:
average = sum(R2_scores) / len(R2_scores)
fold_std = np.std(R2_scores)

print('Average: ', average)
print('Standard Deviation: ', fold_std)

Average:  0.7963781488368554
Standard Deviation:  0.06700513903019562


### 4th block (interpret results)

With the average R^2 score being 0.800, with a standard devation of 0.067, we can say that the data fits our linear regression model well. We may also say that we can use the model to have a better understanding and estimate of the relationship between the duration of a geyser's eruption and the wait time for the next eruption. In otherwords, we can use the duration of the geyser's eruption to estimate how long we would have to wait for the next eruption.